### PR TITLE
Multi- and complex mapping functionality

### PR DESCRIPTION
**Pull request by dschristianson**
Add generalized multi- and complex mapping functionality

- Added new AttributeMapping and MappedAttribure models that will handle basic mapping of all attributes: statistic, result_quality, etc
- Rework plugin mapping file to be general purpose for all datasource to B3D mappings -- Complex mappings handled with : delimiter
- catalog and database structure updated to handle new AttributeMapping model
- Handle all translations automatically in basin3d core translate layer rather than plugins 
  - for query (BASIN-3D vocabulary to datasource vocabulary): the translated query is generated and then validated before attempting to access datasource
  - for model objects (datasource vocabulary to BASIN-3D vocabulary): vocabularies are translated upon object initiation for attributes designated as MappedAttributes
- Streamline vocabularies for result_quality and quality; make translation generalizable in the Result models

ToDo in future:
- Streamline translation in models based on attr typing introspection; see notes in code
- Add a parameter to the catalog.find_attribute_mappings function for “is_compound_mapping” so that only compound mappings are returned. In conjunction: we would either 1) implement a better way to search TinyDB or 2) explore changing to another DB.

Issue #93, #143, #148

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- [x] Unit tests
- [x] Integration tests 
- [x] Test coverage >= 90%
- [x] Flake8 Tests 
- [x] Mypy Tests 

### Test Configuration
* Python Version: 3.9.16

## PR Self Evaluation

- [x] My code follows the agreed upon best practices
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation~
- [x] My changes generate no new warnings
- [x] I have added tests or modified existing tests that prove my fix is effective or that my feature works
- [x] Existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in the appropriate modules
- [x] I have performed a self-review of my own code

### CODE REVIEW

**Comment by dschristianson** [comment]
@vchendrix: Key areas to review:
- catalog.py: new database structure and methods
- plugins.py: get_datasource_mapped_attribute method
- core/synthesis.py: TranslatorMixin and integration into queries
- core/models.py: separate translate_attribute function and internal _translate_attribute methods for models that have mapped attributes.

**Comment by vchendrix** [comment]
@dschristianson I have started reviewing this PR today.  I am going to make comments about small things I find which we may not want to necessarily address in the PR but as new issue tickets.  These are things that have nothing to do with the logic of your changes but the code base as a whole.  

**Integration tests**
Change integration test to check that counts are >= instead of ==.  
```
FAILED tests/integration/test_plugins_usgs.py::test_usgs_monitoring_features[points_by_subbasin] - assert 53 == 52
=================================================================================================== 1 failed, 143 passed, 47 deselected, 1 xfailed in 14.20s =
```

**New errors in newer version of mypy**
I started with a fresh installation of my python environment and discovered that mypy has some new warnings!  See the errors below.  There are only two unique errors and they both relate to new confiruation options

The following are new mypy options that we need to consider
```
# The new default is True. We may just want to set this to False
no_implicit_optional=False

# Either option raise errors
check_untyped_defs=(True|False)
```

### Full mypy error output
```
=========================================================================================================================== FAILURES ===========================================================================================================================
_________________________________________________________________________________________________________________________ test session _________________________________________________________________________________________________________________________
mypy exited with status 1.
_____________________________________________________________________________________________________________________ basin3d/synthesis.py _____________________________________________________________________________________________________________________
51: error: Incompatible default for argument "plugins" (default has type "None", argument has type "List[str]")  [assignment]
51: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
51: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
160: error: Incompatible default for argument "query" (default has type "None", argument has type "Union[QueryById, QueryMonitoringFeature]")  [assignment]
160: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
160: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
245: error: Incompatible default for argument "query" (default has type "None", argument has type "QueryMeasurementTimeseriesTVP")  [assignment]
245: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
245: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
___________________________________________________________________________________________________________________ basin3d/core/catalog.py ____________________________________________________________________________________________________________________
331: error: Incompatible default for argument "variable_names" (default has type "None", argument has type "List[str]")  [assignment]
331: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
331: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
____________________________________________________________________________________________________________________ basin3d/core/models.py ____________________________________________________________________________________________________________________
232: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
233: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
234: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
235: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
236: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
293: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
294: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
350: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
351: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
425: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
426: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
427: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
503: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
504: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
505: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
506: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
507: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
508: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
587: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
662: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
663: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
664: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
665: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
754: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
829: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
830: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
831: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
832: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
833: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
909: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
937: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
938: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1021: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1022: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1023: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1072: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1073: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1074: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1159: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1160: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1161: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1162: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1163: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1164: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1165: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1166: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1294: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1295: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1342: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1343: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1372: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1373: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1403: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1404: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1433: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1434: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1465: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
1466: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
___________________________________________________________________________________________________________________ basin3d/core/monitor.py ____________________________________________________________________________________________________________________
93: error: Incompatible default for argument "name" (default has type "None", argument has type "str")  [assignment]
93: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
93: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
179: error: Incompatible default for argument "log_config_path" (default has type "None", argument has type "str")  [assignment]
179: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
179: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
============================================================================================================================= mypy =============================================================================================================================
Found 6 errors in 3 files (checked 46 source files)

```

**Comment by vchendrix** [comment]
@dschristianson I tried the following code that I got from the [documentation](https://basin3d.readthedocs.io/en/issue-93-multi-mapping/quick_guide.html#get-a-list-of-variables-supported-by-a-basin-3d) to get a list of variables supported by basin3d. 

```
(basin3d) val@vchendrix basin3d % python                              
Python 3.9.13 (main, Oct 13 2022, 16:12:30) 
[Clang 12.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
>>> from basin3d.plugins import usgs
>>> from basin3d import synthesis
>>> synthesizer = synthesis.register()
>>> response = synthesizer.observed_properties()
>>> for opv in response:
...     print(f'{opv.basin3d_vocab} -- {opv.full_name} -- {opv.units}')
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
AttributeError: 'ObservedProperty' object has no attribute 'basin3d_vocab'
>>> next(response)
ObservedProperty(datasource_variable='00060', observed_property_variable=ObservedPropertyVariable(basin3d_id='RDC', full_name='River Discharge', categories=['Hydrogeology', 'Water Physical/Quality Parameters'], units='mV'), sampling_medium='WATER', datasource=DataSource(id='USGS', name='USGS', id_prefix='USGS', location='https://waterservices.usgs.gov/nwis/', credentials={}), datasource_description='Discharge, cubic feet per second')

```

**Comment by dschristianson** [comment]
@vchendrix: That is the previous ObservedProperty and ObservedPropertyVariable models. The former is different now and ObservedPropertyVariable does not even exist anymore. I've double checked the code. Is something amiss in your setup?

**Comment by vchendrix** [comment]
> # Either option raise errors
> check_untyped_defs=(True|False)

Ignore this one. in my effort to create a new environment for testing I forgot to check out the pr branch.  However the other mypy config options still needs to be addressed.

**Comment by vchendrix** [comment]
@dschristianson when instead of deleting mapping_usgs.csv and adding a new usgs_mapping.csv use `git mv` to rename the file so we can see the diffs.

**Comment by vchendrix** [comment]
> @dschristianson when instead of deleting mapping_usgs.csv and adding a new usgs_mapping.csv use `git mv` to rename the file so we can see the diffs.

same with mapping files for the test plugins

**Comment by vchendrix** [comment]
> > @dschristianson when instead of deleting mapping_usgs.csv and adding a new usgs_mapping.csv use `git mv` to rename the file so we can see the diffs.
> 
> same with mapping files for the test plugins

Also, I am curious why you changed the format of the naming.  I think I understand why, so it sorts with the usgs plugin which I think am onboard with and see the value in for organizational purposes.  It was the reverse to may the type prominent but I see now that that has little value.  :)

File: basin3d/core/schema/query.py [hunk 1]
```diff
@@ -95,21 +105,29 @@ def __init__(self, **data):
                 data[field] = isinstance(data[field], str) and data[field].upper() or data[field]
         super().__init__(**data)
 
+    def get_prefixed_fields(self) -> list:
+        return ['monitoring_feature', 'parent_feature']
+
```
Comment (vchendrix) [inline basin3d/core/schema/query.py:110]: This should be a field that is a pydantic ClassVar instead

```suggestion
   prefixed_fields: ClassVar[List[str]] = ['monitoring_feature', 'parent_feature']

```

File: basin3d/core/schema/query.py [hunk 2]
```diff
@@ -61,6 +63,14 @@ class Config:
...
         validate_all = True
 
+    def get_mapped_fields(self) -> list:
+        """Get the query fields that have mappings. Subclasses may overwrite this base function"""
+        return []
+
+    def get_prefixed_fields(self) -> list:
+        """Get the query fields that have prefixes. Subclasses may overwrite ths base function"""
+        return []
+
```
Comment (vchendrix) [inline basin3d/core/schema/query.py:73]: This should be a field that is a pydantic ClassVar instead

```suggestion
   prefixed_fields: ClassVar[List[str]] = []

```

File: basin3d/core/schema/query.py [hunk 3]
```diff
@@ -118,13 +136,34 @@ def __init__(self, **data):
...
+        return data
+
+    def get_mapped_fields(self) -> list:
+        # observed_property_variables is first b/c it is most likely to have compound mappings.
+        # ToDo: check how order may affect translation (see core/synthesis)
+        return ['observed_property', 'aggregation_duration', 'statistic', 'result_quality', 'sampling_medium']
+
+    def get_prefixed_fields(self) -> list:
+        return ['monitoring_feature']
+
```
Comment (vchendrix) [inline basin3d/core/schema/query.py:166]: This should be a ClassVar field (from pydantic)
```suggestion
    prefixed_fields: ClassVar[List[str]] = ['monitoring_feature']

```

**Review by vchendrix — CHANGES_REQUESTED** [review]
@dschristianson I have not walked through the documentation and tried the examples yet but I wanted to get the review to your sooner than later.  Overall I think it looks great!  There are just a few suggestions: documentation on some function args, a small change to how you are defining prefixed fields in the schemas.

I will try to review the documentation tomorrow or Thursday.  Great job!

File: basin3d/core/models.py [hunk 4]
```diff
@@ -108,32 +135,89 @@ def __str__(self):
...
+    A data class for an attribute that is translated (i.e., mapped) from a datasource vocabulary to BASIN-3D vocabulary.
+    Note that this model holds an AttributeMapping that maybe compound in nature; however this class specifies only one attribute types.
+    For example, if the AttributeMapping is for a compound mapping of attribute types OBSERVED_PROPERTY:SAMPLING_MEDIUM,
+    then the attr_type field would be either OBSERVED_PROPERTY or SAMPLING_MEDIUM but not both.
+
+    Fields:
+         - *attr_type:* Attribute Type; e.g., STATISTIC, RESULT_QUALITY, OBSERVED_PROPERTY, etc; single type only
+         - *attr_mapping:* AttributeMapping as described in the datasource's (i.e., plugin's mapping file).
+    """
+    attr_type: str
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/models.py:159]: Yes, can be MappedAttributeEnum. Changed.

File: basin3d/core/plugin.py [hunk 5]
```diff
@@ -63,63 +64,74 @@ def __init__(self, datasource: DataSource, catalog: CatalogTinyDb):
...
+        Get other attributes if attr_type is part of compound mapping
+
+        :param attr_type:
+        :param is_query:
+        :return:
         """
+        return self._catalog.find_compound_mapping_attributes(self.datasource.id, attr_type, is_query=is_query)
 
-        return self._catalog.find_observed_property_variable(self.datasource.id, variable_name, from_basin3d)
+    def get_compound_mapping_str(self):
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/plugin.py]: This method no longer exists. Not sure why you were still seeing it??

File: basin3d/core/catalog.py [hunk 6]
```diff
@@ -100,22 +105,32 @@ def _insert(self, record):
...
-        Access a single observed property
+        Access a single attribute mapping
 
-        :param datasource_id:  datasource identifier
-        :param basin3d_id:  BASIN-3D variable identifier
-        :param datasource_variable_id: datasource variable identifier
+        :param datasource_id:
+        :param attr_type:
+        :param basin3d_vocab:
+        :param datasource_vocab:
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/catalog.py:133]: Question about this: I put the documentation in the TinyDB class. I wasn't sure if I should repeat it in the base class or if it should only go in one? Putting it in both means maintaining in both, while putting it only in the base class seemed wrong also b/c it is far from the method doing the work.

I'll go ahead in put it in both places to address this comment and the other similar ones. Look forward to your further thoughts.

File: basin3d/core/catalog.py [hunk 7]
```diff
@@ -124,68 +139,51 @@ def is_initialized(self) -> bool:
...
-    def find_observed_properties(self, datasource_id=None, variable_names=None) -> Iterator[ObservedProperty]:
+    def find_observed_properties(self, basin3d_vocab=None) -> Iterator[Optional[ObservedProperty]]:
         """
-        Get the measurement to the specified variable_name
 
-        :param datasource_id:  the datasource identifier
-        :param variable_names: the variable names to get the :class:`~basin3d.models.ObservedProperty` for
-        :type variable_names: list
-        :return: :class:`~basin3d.models.ObservedProperty`
+        :param basin3d_vocab:
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/catalog.py:153]: description copied from TinyDB class

File: basin3d/core/catalog.py [hunk 8]
```diff
@@ -124,68 +139,51 @@ def is_initialized(self) -> bool:
...
-        :param variable_name:  The :class:`~basin3d.models.ObservedPropertyVariable`
-             name to convert
-        :param: from_basin3d: boolean that says whether the variable name is a
-            BASIN-3D variable. If not, then this a datasource variable name.
-        :type from_basin3d: boolean
-        :return: A variable name
-        :rtype: str
+        :param datasource_id:
+        :param attr_type:
+        :param attr_vocab:
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/catalog.py:163]: description copied from TinyDB class

File: basin3d/core/catalog.py [hunk 9]
```diff
@@ -124,68 +139,51 @@ def is_initialized(self) -> bool:
...
-        :type variable_names: iterable
-        :param: from_basin3d: boolean that says whether the variable name is a
-            BASIN-3D variable. If not, then this a datasource variable names.
-        :type from_basin3d: boolean
-        :return: list of variable names
-        :rtype: iterable
+        :param datasource:
+        :param attr_type:
+        :param attr_vocab:
+        :param from_basin3d:
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/catalog.py:174]: description copied from TinyDB class

File: basin3d/plugins/usgs.py [hunk 10]
```diff
@@ -657,10 +619,10 @@ def list(self, query: QueryMeasurementTimeseriesTVP):
...
                 monitoring_feature = None
 
-            # deal with statistic
-            basin3d_statistic = "NOT_SET"
-            if statistic:
-                basin3d_statistic = map_statistic_code(statistic)
+            # # deal with statistic
+            # basin3d_statistic = "NOT_SET"
+            # if statistic:
+            #     basin3d_statistic = map_statistic_code(statistic)
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/plugins/usgs.py:625]: done. good catch!

File: basin3d/core/catalog.py [hunk 11]
```diff
@@ -236,235 +247,386 @@ def _process_plugin_variable_mapping(self, plugin, map_filename: str, datasource
...
+
+        if datasource_id and self._get_datasource(datasource_id) is None:
+            msg = f'Specified datasource_id: "{datasource_id}" has not been registered.'
+            logger.critical(msg)
+            raise CatalogException(msg)
+
+        if attr_type and attr_type not in MappedAttributeEnum.values():
+            msg = f'"{attr_type}" is not an attribute type supported by BASIN-3D.'
+            logger.critical(msg)
+            raise CatalogException(msg)
```
[the remark this answers is not in the document]
Comment (dschristianson) [reply basin3d/core/catalog.py:468]: implemented in parent class.

File: basin3d/core/schema/query.py (hunk 1 above)
Comment (dschristianson) [reply basin3d/core/schema/query.py:110]: OK. Done.

File: basin3d/core/schema/query.py (hunk 2 above)
Comment (dschristianson) [reply basin3d/core/schema/query.py:73]: Done. Changed get_mapped_fields similarly.

File: basin3d/core/schema/query.py (hunk 3 above)
Comment (dschristianson) [reply basin3d/core/schema/query.py:166]: Done. Changed get_mapped_fields similarly.

**Review by vchendrix — APPROVED** [review]
@dschristianson I am going to approve this.  I feel that the documentation review can happen as a part of the final release I will be doing when I review the documentation as a whole.  There are two commits in her that have no issue number associated with them. I will be squashing the commit with no issue numbers in with the main commit.

**Review by vchendrix — APPROVED** [review]

**Pull request merged**